port/unix/omrsysinfo.c

```diff
@@ -4315,6 +4380,85 @@ omrsysinfo_is_running_in_container(struct OMRPortLibrary *portLibrary, BOOLEAN *
 	return rc;
 }
 
+int32_t
+omrsysinfo_cgroup_subsystem_iterator_init(struct OMRPortLibrary *portLibrary, uint64_t subsystem, struct OMRCgroupMetricIteratorState *state)
+{
+	Assert_PRT_true(NULL != state);
+	int32_t rc = OMRPORT_ERROR_SYSINFO_CGROUP_UNSUPPORTED_PLATFORM;
+#if defined(LINUX) && !defined(OMRZTPF)
+	state->count = 0;
+	switch (subsystem) {
+		case OMR_CGROUP_SUBSYSTEM_MEMORY :
+			 state->numElements = sizeof(omrCgroupMemoryMetricMap) / sizeof(omrCgroupMemoryMetricMap[0]);
+			 state->subsystemid = OMR_CGROUP_SUBSYSTEM_MEMORY;
+			 break;
+		case OMR_CGROUP_SUBSYSTEM_CPU :
+			 state->numElements = sizeof(omrCgroupCpuMetricMap) / sizeof(omrCgroupCpuMetricMap[0]);
+			 state->subsystemid = OMR_CGROUP_SUBSYSTEM_CPU;
+			 break;
+		case OMR_CGROUP_SUBSYSTEM_CPUSET :
+			 state->numElements = sizeof(omrCgroupCpusetMetricMap) / sizeof(omrCgroupCpusetMetricMap[0]);
+			 state->subsystemid = OMR_CGROUP_SUBSYSTEM_CPUSET;
+			 break;
+		default :
+			goto _end;
+	}
+	rc = 0;
+#endif /* defined(LINUX) && !defined(OMRZTPF) */
+_end:
+	return rc;
+}
+
+BOOLEAN
+omrsysinfo_cgroup_subsystem_iterator_hasNext(struct OMRPortLibrary *portLibrary, const struct OMRCgroupMetricIteratorState *state)
+{
+	BOOLEAN check = FALSE;
+#if defined(LINUX) && !defined(OMRZTPF)
+	check = state->count < state->numElements;
+#endif /* defined(LINUX) && !defined(OMRZTPF) */
+	return check;
+}
+
+int32_t
+omrsysinfo_cgroup_subsystem_iterator_next(struct OMRPortLibrary *portLibrary, struct OMRCgroupMetricIteratorState *state, struct OMRCgroupMetricElement *metricElement, BOOLEAN *printUnits)
+{
+	int32_t rc = OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_UNAVAILABLE;
+#if defined(LINUX) && !defined(OMRZTPF)
+	const struct OMRCgroupSubsystemMetricMap *subsystemMetricMap = NULL;
+	if (state->count >= state->numElements) {
+		goto _end;
+	}
+	switch (state->subsystemid) {
+		case OMR_CGROUP_SUBSYSTEM_MEMORY :
+			subsystemMetricMap = omrCgroupMemoryMetricMap;
+			break;
+		case OMR_CGROUP_SUBSYSTEM_CPU :
+			subsystemMetricMap = omrCgroupCpuMetricMap;
+			break;
+		case OMR_CGROUP_SUBSYSTEM_CPUSET :
+			subsystemMetricMap = omrCgroupCpusetMetricMap;
+			break;
+		default:
+			rc = OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_UNAVAILABLE;
+			goto _end;
+	}
+	metricElement->key = subsystemMetricMap[state->count].metricTag;
+	metricElement->units = subsystemMetricMap[state->count].metricUnit;
+	rc = readCgroupMetricFromFile(portLibrary, state->subsystemid, subsystemMetricMap[state->count].metricFilename, subsystemMetricMap[state->count].metricKeyInFile, metricElement->value);
```

Comment:
this opens the file on each call, correct?  Can we find a better way to do this?

Reply:
@DanHeidinga after going through your review i thought that the current mechanism is not an optimised one so would like to reshape it to a newer one which is described below:

With few modifications in the 4 parts of code (init, hasnext, next, structures) i guess it could clear out most of the issues like re-reading and searching.

**Pseudocode**


File with multiple metrics is read into a buffer and kept in a struct with filename and content

make structs holding all the required info for the metric element
Have a structure for each multiline content we need
loop every subsystem

if its multiline (can get the count from the array so if its >1 (SINGLE_CGROUP_METRIC) then its multiline)
```
    mark the secoundary counter in state to length of metric struct array
    read the file and store in a buffer (store to a struct with content and file name)
    process the first element and return and keep a count in state (secondary counter in state).
    donot increment the state count so that hasNext drives again to here
    process the element from buffer on every iteration till the secondary counter becomes 0
    at end of metric array  free up the memory of struct (filename, content buffer) 
```
else
```
    read the file to value buffer but not to the storage buffer  
```

**Sample structure format**
```
#define SINGLE_CGROUP_METRIC 1

typedef struct OMRCgroupMetricInfoElement {
    char *metricTag;
    char *metricKeyInFile;
    char *metricUnit;
    BOOLEAN isValueToBeChecked;
} OMRCgroupMetricInfoElement;

typedef struct OMRCgroupMetricFileContent {
     char *filename;
     char contentBuffer[1024];
} OMRCgroupMetricFileContent;

struct OMRCgroupMetricInfoElement oomControlMetricElementList[] = {
    { "OOM Killer Disabled", "oom_kill_disable", "", FALSE },
    { "Under OOM", "under_oom", "", FALSE }
};

struct OMRCgroupMetricInfoElement cpuStatMetricElementList[] = {
    { "Period intervals elapsed count", "nr_periods", "", FALSE },
    { "Throttled count", "nr_throttled", "", FALSE },
    { "Total throttle time", "throttled_time", "nanoseconds", FALSE }
};

typedef struct OMRCgroupSubsystemMetricMap {
    char *metricFileName;
    OMRCgroupMetricInfoElement *metricInfoElementList;
    int32_t metricElementsCount;
} OMRCgroupSubsystemMetricMap;

struct OMRCgroupSubsystemMetricMap omrCgroupMemoryMetricMap[] = {
    { "/sys/fs/cgroup/memory/memory.limit_in_bytes", &(OMRCgroupMetricInfoElement){ "Memory Limit", NULL, "bytes", TRUE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes", &(OMRCgroupMetricInfoElement){ "Memory + Swap Limit", NULL, "bytes", TRUE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.usage_in_bytes", &(OMRCgroupMetricInfoElement){ "Memory Usage", NULL, "bytes", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.memsw.usage_in_bytes", &(OMRCgroupMetricInfoElement){ "Memory + Swap Usage", NULL, "bytes", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.max_usage_in_bytes", &(OMRCgroupMetricInfoElement){ "Memory Max Usage", NULL, "bytes", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.memsw.max_usage_in_bytes", &(OMRCgroupMetricInfoElement){ "Memory + Swap Max Usage", NULL, "bytes", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.failcnt", &(OMRCgroupMetricInfoElement){ "Memory limit exceeded count", NULL, "", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.memsw.failcnt", &(OMRCgroupMetricInfoElement){ "Memory + Swap limit exceeded count", NULL, "", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/memory/memory.oom_control", &oomControlMetricElementList, sizeof(oomControlMetricElementList)/sizeof(oomControlMetricElementList[0]) }
};

struct OMRCgroupSubsystemMetricMap omrCgroupCpuMetricMap[] = {
    { "/sys/fs/cgroup/cpu/cpu.cfs_period_us", &(OMRCgroupMetricInfoElement){ "CPU Period", NULL, "microseconds", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/cpu/cpu.cfs_quota_us", &(OMRCgroupMetricInfoElement){ "CPU Quota", NULL, "microseconds", TRUE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/cpu/cpu.shares", &(OMRCgroupMetricInfoElement){ "CPU Shares", NULL, "", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/cpu/cpu.stat", &cpuStatMetricElementList, sizeof(cpuStatMetricElementList)/sizeof(cpuStatMetricElementList[0]) }
};

struct OMRCgroupSubsystemMetricMap omrCgroupCpusetMetricMap[] = {
    { "/sys/fs/cgroup/cpuset/cpuset.cpu_exclusive", &(OMRCgroupMetricInfoElement){ "CPU exclusive", NULL, "", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/cpuset/cpuset.mem_exclusive", &(OMRCgroupMetricInfoElement){ "Mem exclusive", NULL, "", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/cpuset/cpuset.cpus", &(OMRCgroupMetricInfoElement){ "CPUs", NULL, "", FALSE }, SINGLE_CGROUP_METRIC },
    { "/sys/fs/cgroup/cpuset/cpuset.mems", &(OMRCgroupMetricInfoElement){ "Mems", NULL, "", FALSE }, SINGLE_CGROUP_METRIC }
};
```